File: src/main/java/com/lk/setl/util/StringUtils.java
package com.lk.setl.util;

import java.util.regex.Pattern;

public class StringUtils {

    /**
     * Given a hex string this will return the byte array corresponding to the string .
     *
     * @param hex the hex String array
     * @return a byte array that is a hex string representation of the given string. The size of the
     *     byte array is therefore hex.length/2
     */
    public static byte[] hexStringToByte(final String hex) {
        final byte[] bts = new byte[hex.length() / 2];
        for (int i = 0; i < bts.length; i++) {
            bts[i] = (byte) Integer.parseInt(hex.substring(2 * i, 2 * i + 2), 16);
        }
        return bts;
    }

    public static String escapeLikeRegex(String pattern, char escapeChar) {
        StringBuilder out = new StringBuilder();
        int length = pattern.length();

        for (int i = 0; i < length; i++) {
            char c = pattern.charAt(i);
            if (c == escapeChar && i + 1 < length) {
                i++;
                c = pattern.charAt(i);
                if (c == '_' || c == '%') {
                    out.append(Pattern.quote(Character.toString(c)));
                } else if (c == escapeChar) {
                    out.append(Pattern.quote(Character.toString(c)));
                } else {
                    throw new IllegalArgumentException(String.format("the escape character is not allowed to precede '%s'", Character.toString(c)));
                }
            } else if (c == escapeChar) {
                throw new IllegalArgumentException("it is not allowed to end with the escape character");
            } else if (c == '_') {
                out.append(".");
            } else if (c == '%') {
                out.append(".*");
            } else {
                out.append(Pattern.quote(Character.toString(c)));
            }
        }

        return "(?s)" + out.toString(); // (?s) enables dotall mode, causing "." to match new lines
    }

    public static String substringSQL(String string, int pos, int len){
        if(string == null){
            return null;
        }

        int length = string.length();
        if (pos > length) {
            return "";
        }

        int start = 0;
        int end;
        if (pos > 0) {
            start = pos - 1;
        } else if (pos < 0) {
            start = length + pos;
        }
        if ((length - start) < len) {
            end = length;
        } else {
            end = start + len;
        }
        start = Math.max(start, 0); // underflow
        if (start >= end) {
            return "";
        }

        return  string.substring(start, end);
    }

    public static String trim(String srcStr, boolean leading, boolean trailing, String trimStr) {
        trimStr = trimStr == null || trimStr.isEmpty() ? " " : trimStr;
        int begin = 0, end = srcStr.length();
        if (leading) {
            while (begin < end && trimStr.indexOf(srcStr.charAt(begin)) >= 0) {
                begin++;
            }
        }
        if (trailing) {
            while (end > begin && trimStr.indexOf(srcStr.charAt(end - 1)) >= 0) {
                end--;
            }
        }
        // substring() returns self if start == 0 && end == length()
        return srcStr.substring(begin, end);
    }
}

File: src/main/java/com/lk/setl/sql/catalyst/util/StringUtils.scala
package com.lk.setl.sql.catalyst.util

import com.lk.setl.Logging
import com.lk.setl.sql.AnalysisException

import java.util.regex.{Pattern, PatternSyntaxException}
import scala.collection.mutable.ArrayBuffer

object StringUtils extends Logging {

  /**
   * Validate and convert SQL 'like' pattern to a Java regular expression.
   *
   * Underscores (_) are converted to '.' and percent signs (%) are converted to '.*', other
   * characters are quoted literally. Escaping is done according to the rules specified in
   * [[Like]] usage documentation. An invalid pattern will
   * throw an [[AnalysisException]].
   *
   * @param pattern the SQL pattern to convert
   * @param escapeChar the escape string contains one character.
   * @return the equivalent Java regular expression of the pattern
   */
  def escapeLikeRegex(pattern: String, escapeChar: Char): String = {
    val in = pattern.toIterator
    val out = new StringBuilder()

    def fail(message: String) = throw new AnalysisException(
      s"the pattern '$pattern' is invalid, $message")

    while (in.hasNext) {
      in.next match {
        case c1 if c1 == escapeChar && in.hasNext =>
          val c = in.next
          c match {
            case '_' | '%' => out ++= Pattern.quote(Character.toString(c))
            case c if c == escapeChar => out ++= Pattern.quote(Character.toString(c))
            case _ => fail(s"the escape character is not allowed to precede '$c'")
          }
        case c if c == escapeChar => fail("it is not allowed to end with the escape character")
        case '_' => out ++= "."
        case '%' => out ++= ".*"
        case c => out ++= Pattern.quote(Character.toString(c))
      }
    }
    "(?s)" + out.result() // (?s) enables dotall mode, causing "." to match new lines
  }

  private[this] val trueStrings =
    Set("t", "true", "y", "yes", "1")

  private[this] val falseStrings =
    Set("f", "false", "n", "no", "0")

  // scalastyle:off caselocale
  def isTrueString(s: String): Boolean = trueStrings.contains(s.strip().toLowerCase)

  def isFalseString(s: String): Boolean = falseStrings.contains(s.strip().toLowerCase)
  // scalastyle:on caselocale

  /**
   * This utility can be used for filtering pattern in the "Like" of "Show Tables / Functions" DDL
   * @param names the names list to be filtered
   * @param pattern the filter pattern, only '*' and '|' are allowed as wildcards, others will
   *                follow regular expression convention, case insensitive match and white spaces
   *                on both ends will be ignored
   * @return the filtered names list in order
   */
  def filterPattern(names: Seq[String], pattern: String): Seq[String] = {
    val funcNames = scala.collection.mutable.SortedSet.empty[String]
    pattern.trim().split("\\|").foreach { subPattern =>
      try {
        val regex = ("(?i)" + subPattern.replaceAll("\\*", ".*")).r
        funcNames ++= names.filter{ name => regex.pattern.matcher(name).matches() }
      } catch {
        case _: PatternSyntaxException =>
      }
    }
    funcNames.toSeq
  }

  /**
   * Concatenation of sequence of strings to final string with cheap append method
   * and one memory allocation for the final string.  Can also bound the final size of
   * the string.
   */
  class StringConcat(val maxLength: Int = 100000000) {
    protected val strings = new ArrayBuffer[String]
    protected var length: Int = 0

    def atLimit: Boolean = length >= maxLength

    /**
     * Appends a string and accumulates its length to allocate a string buffer for all
     * appended strings once in the toString method.  Returns true if the string still
     * has room for further appends before it hits its max limit.
     */
    def append(s: String): Unit = {
      if (s != null) {
        val sLen = s.length
        if (!atLimit) {
          val available = maxLength - length
          val stringToAppend = if (available >= sLen) s else s.substring(0, available)
          strings.append(stringToAppend)
        }

        // Keeps the total length of appended strings. Note that we need to cap the length at
        // `ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH`; otherwise, we will overflow
        // length causing StringIndexOutOfBoundsException in the substring call above.
        length = Math.min(length.toLong + sLen, maxLength).toInt
      }
    }

    /**
     * The method allocates memory for all appended strings, writes them to the memory and
     * returns concatenated string.
     */
    override def toString: String = {
      val finalLength = if (atLimit) maxLength else length
      val result = new java.lang.StringBuilder(finalLength)
      strings.foreach(result.append)
      result.toString
    }
  }

  /**
   * A string concatenator for plan strings.  Uses length from a configured value, and
   *  prints a warning the first time a plan is truncated.
   */
  class PlanStringConcat extends StringConcat(Math.max(0, 10000000)) {
    override def toString: String = {
      if (atLimit) {
        logWarning(
          "Truncated the string representation of a plan since it was too long. The " +
            s"plan had length ${length} and the maximum is ${maxLength}. This behavior " +
            s"can be adjusted by setting 'MAX_PLAN_STRING_LENGTH'.")
        val truncateMsg = if (maxLength == 0) {
          s"Truncated plan of $length characters"
        } else {
          s"... ${length - maxLength} more characters"
        }
        val result = new java.lang.StringBuilder(maxLength + truncateMsg.length)
        strings.foreach(result.append)
        result.append(truncateMsg)
        result.toString
      } else {
        super.toString
      }
    }
  }
}
